src/Microsoft.IdentityModel.Tokens/Encryption/KeyWrapProvider.cs

```diff
@@ -0,0 +1,439 @@
+        private ICryptoTransform _symmetricAlgorithmEncryptor;
+        private ICryptoTransform _symmetricAlgorithmDecryptor;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="KeyWrapProvider"/> class used for wrap key and unwrap key.
+        /// <param name="key">The <see cref="SecurityKey"/> that will be used for crypto operations.</param>
+        /// <param name="algorithm">The KeyWrap algorithm to apply.</param>
+        /// <exception cref="ArgumentNullException">'key' is null.</exception>
+        /// <exception cref="ArgumentNullException">'algorithm' is null.</exception>
+        /// <exception cref="ArgumentException">If <see cref="SecurityKey"/> and algorithm pair are not supported.</exception>
+        /// <exception cref="InvalidOperationException"><see cref="KeyWrapProvider.GetSymmetricAlgorithm"/> throws.</exception>
+        /// </summary>
+        public KeyWrapProvider(SecurityKey key, string algorithm)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            if (string.IsNullOrEmpty(algorithm))
+                throw LogHelper.LogArgumentNullException(nameof(algorithm));
+
+            if (!IsSupportedAlgorithm(key, algorithm))
+                throw LogHelper.LogExceptionMessage(new ArgumentException(String.Format(CultureInfo.InvariantCulture, LogMessages.IDX10661, algorithm, key)));
+
+            Algorithm = algorithm;
+            Key = key;
+
+            _symmetricAlgorithm = GetSymmetricAlgorithm();
+        }
+
+        /// <summary>
+        /// Returns the <see cref="SymmetricAlgorithm"/>.
+        /// </summary>
+        /// <returns></returns>
+        /// <exception cref="ArgumentException">The <see cref="SecurityKey"/> cannot be converted to byte array</exception>
+        /// <exception cref="ArgumentOutOfRangeException">The keysize doesn't match the algorithm.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">The algorithm doesn't support algorithm.</exception>
+        /// <exception cref="InvalidOperationException">Failed to create symmetric algorithm with provided key and algorithm.</exception>
+        protected virtual SymmetricAlgorithm GetSymmetricAlgorithm()
+        {
+            byte[] keyBytes = null;
+
+            SymmetricSecurityKey symmetricSecurityKey = Key as SymmetricSecurityKey;
+            if (symmetricSecurityKey != null)
+                keyBytes = symmetricSecurityKey.Key;
+            else
+            {
+                JsonWebKey jsonWebKey = Key as JsonWebKey;
+                if (jsonWebKey != null && jsonWebKey.K != null && jsonWebKey.Kty == JsonWebAlgorithmsKeyTypes.Octet)
+                    keyBytes = Base64UrlEncoder.DecodeBytes(jsonWebKey.K);
+            }
+
+            if (keyBytes == null)
+                throw LogHelper.LogExceptionMessage(new ArgumentException(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10657, Key.GetType())));
```

Comment:
You shouldn't throw an InvalidCastException- it gets thrown if you cast something improperly. Please make this an InvalidOperationException.